File: 1671-minimum-number-of-removals-to-make-mountain-array/1671-minimum-number-of-removals-to-make-mountain-array.java
class Solution {
    public int minimumMountainRemovals(int[] nums) {
        int n=nums.length;
        
        int[] dps=new int[n];
        int[] dpe=new int[n];
        
        for(int i=0; i<n; i++){
            dps[i]=1;
            for(int j=i-1; j>=0; j--){
                if(nums[i]>nums[j]){
                    dps[i]=Math.max(dps[i],dps[j]+1);
                }
            }
        }
        
        for(int i=n-1; i>=0; i--){
            dpe[i]=1;
            for(int j=i+1; j<n; j++){
                if(nums[i]>nums[j]){
                    dpe[i]=Math.max(dpe[i],dpe[j]+1);
                }
            }
        }
        
        int lbs=0;
        for(int i=1; i<n-1; i++){
            if(dps[i]==1) continue;
            if(dpe[i]==1) continue;
            lbs=Math.max(lbs,dps[i]+dpe[i]-1);
        }
        
        return n-lbs;
    }
}

File: 1671-minimum-number-of-removals-to-make-mountain-array/1671-minimum-number-of-removals-to-make-mountain-array.cpp
class Solution {
public:
    int minimumMountainRemovals(vector<int>& nums) {
         int n=nums.size();

        vector<int> dp_lr(n,0);


        for(int i=0; i<n; i++){
            dp_lr[i]=1;

            for(int j=i-1; j>=0; j--){
                if(nums[j]<nums[i]){
                    dp_lr[i]=max(dp_lr[i],dp_lr[j]+1);
                }
            }
        }

        vector<int> dp_rl(n,0);

        for(int i=n-1; i>=0; i--){
            dp_rl[i]=1;

            for(int j=i+1; j<n; j++){
                if(nums[j]<nums[i]){
                    dp_rl[i]=max(dp_rl[i],dp_rl[j]+1);
                }
            }
        }

        int ans=0;

        for(int i=1; i<n-1; i++){
            if(dp_lr[i]==1 || dp_rl[i]==1){
                continue;
            }
            
            ans=max(ans,dp_lr[i]+dp_rl[i]-1);
        }

        return n-ans;
    }
};